python/metatensor-learn/tests/module_map.py

```diff
@@ -87,3 +93,126 @@ def test_module_map_single_block_tensor(single_block_tensor_torch, out_propertie
+
+    # the input keys and output properties are still on cpu, because
+    # metatensor.Labels are never moved to a different device
+    assert tensor_module._in_keys.device == "cpu"
+    if out_properties is not None:
+        for label in tensor_module._out_properties:
+            assert label.device == "cpu"
+
+
+@pytest.mark.parametrize(
+    "out_properties", [None, [Labels(["a", "b"], np.array([[1, 1]]))]]
+)
+@pytest.mark.skipif(not HAS_CUDA, reason="requires cuda")
+@pytest.mark.filterwarnings(
+    "ignore:.*If you are using Pytorch and need the labels to also be on cuda.*"
+)
+def test_module_map_cuda(single_block_tensor_torch, out_properties):  # noqa F811
+    """
+    Checks the `to` function of module map by moving the module to cuda and checking
+    that the output tensor is on cuda device.
+    """
+    modules = []
+    for key in single_block_tensor_torch.keys:
+        modules.append(
+            MockModule(
+                in_features=len(single_block_tensor_torch.block(key).properties),
+                out_features=5,
+            )
+        )
+    tensor_module = ModuleMap(
+        single_block_tensor_torch.keys, modules, out_properties=out_properties
+    )
+
+    assert tensor_module._in_keys.device == "cpu"
+    if out_properties is not None:
+        for label in tensor_module._out_properties:
+            assert label.device == "cpu"
+
+    tensor_module.to("cuda")
+
+    # at this point, the parameters should have been moved,
+    # but the input keys and output properties should still be on cpu
+    for parameter in tensor_module.parameters():
+        assert parameter.device.type == "cuda"
+
+    assert tensor_module._in_keys.device == "cpu"
+    if out_properties is not None:
+        for label in tensor_module._out_properties:
+            assert label.device == "cpu"
+
+    single_block_tensor_torch = single_block_tensor_torch.to(device="cuda")
+    out_tensor = tensor_module(single_block_tensor_torch)
+    assert out_tensor.device.type == "cuda"
+
+    # the input keys and output properties are still on cpu, because
+    # metatensor.Labels are never moved to a different device
+    assert tensor_module._in_keys.device == "cpu"
+    if out_properties is not None:
+        for label in tensor_module._out_properties:
+            assert label.device == "cpu"
+
+
+def test_module_map_dtype(single_block_tensor_torch):
+    modules = []
+    for key in single_block_tensor_torch.keys:
+        modules.append(
+            MockModule(
+                in_features=len(single_block_tensor_torch.block(key).properties),
+                out_features=5,
+            )
+        )
+
+    tensor_module = ModuleMap(single_block_tensor_torch.keys, modules)
+    tensor_module(single_block_tensor_torch)
+
+    tensor_module = tensor_module.to(torch.float64)
+    single_block_tensor_torch = single_block_tensor_torch.to(dtype=torch.float64)
+    tensor_module(single_block_tensor_torch)
```

Comment:
please add an assert that the output has the right dtype before and after the `to` call